### PR TITLE
Rush incremental: respect command line flags

### PR DESCRIPTION
**Pull request by nickpape-msft**
Right now, Rush build won't do anything even if the command line flags are changing. This adds the build command into the package-deps file, which ensures that if the build command changes that we also trigger a full rebuild.

### CODE REVIEW

File: apps/rush/src/taskRunner/ProjectBuildTask.ts [hunk 1]
```diff
@@ -67,44 +76,59 @@ export default class ProjectBuildTask implements ITaskDefinition {
...
-    let lastPackageDeps: IPackageDeps;
+    let lastPackageDeps: IPackageDependencies;
 
     try {
       writer.writeLine(`>>> ${this.name}`);
 
+      const currentDepsPath: string = path.join(this._rushProject.projectFolder, RushConstants.packageDepsFilename);
       if (fsx.existsSync(currentDepsPath)) {
-        lastPackageDeps = JSON.parse(fsx.readFileSync(currentDepsPath, 'utf8')) as IPackageDeps;
+        lastPackageDeps = JSON.parse(fsx.readFileSync(currentDepsPath, 'utf8')) as IPackageDependencies;
```
Comment (pgonzal) [inline apps/rush/src/taskRunner/ProjectBuildTask.ts]: >JSON.parse [](start = 26, length = 10)

Use JsonFile.readJsonFile(), which uses JJU to give professional quality error messages.

**Comment by pgonzal** [comment]
>             fsx.writeFileSync(currentDepsPath, JSON.stringify(currentPackageDeps, undefined, 2));

Use JsonFile.writeJsonFile()

---
Refers to: apps/rush/src/taskRunner/ProjectBuildTask.ts:212 in e5d7a7c. [](commit_id = e5d7a7c78a24629ed1560d9174cdbb63f8c1be1b, deletion_comment = False)

File: apps/rush/src/taskRunner/ProjectBuildTask.ts [hunk 2]
```diff
@@ -151,22 +168,8 @@ export default class ProjectBuildTask implements ITaskDefinition {
...
-          build.args.push('--npm');
-        }
-        if (this._minimalMode) {
-          build.args.push('--minimal');
-        }
-
-        const actualBuildCommand: string = `${build.command} ${build.args.join(' ')}`;
-
         // Run the test step
+        const actualBuildCommand: string = `${build.command} ${build.args.join(' ')}`;
```
Comment (pgonzal) [inline apps/rush/src/taskRunner/ProjectBuildTask.ts]: >{build.args [](start = 62, length = 11)

This is not escaped properly.

**Comment by pgonzal** [comment]
:clock1:

File: apps/rush/src/taskRunner/ProjectBuildTask.ts (hunk 2 above)
Comment (nickpape-msft) [inline apps/rush/src/taskRunner/ProjectBuildTask.ts]: it has been working so far... how should it be escaped?

---
In reply to: [116881577](https://github.com/Microsoft/web-build-tools/pull/195#discussion_r116881577) [](ancestors = 116881577)

**Comment by nickpape-msft** [comment]
@qz2017 is added to the review. #Closed

**Comment by nickpape-msft** [comment]
@iclanton is added to the review. #Closed

File: apps/rush/src/taskRunner/ProjectBuildTask.ts [hunk 3]
```diff
@@ -221,7 +217,32 @@ export default class ProjectBuildTask implements ITaskDefinition {
...
+
+    if (this._production) {
+      build.args.push('--production');
+    }
+    if (this._npmMode) {
+      build.args.push('--npm');
+    }
+    if (this._minimalMode) {
+      build.args.push('--minimal');
+    }
```
Comment (iclanton) [inline apps/rush/src/taskRunner/ProjectBuildTask.ts]: Ugh we should work on deprecating these soon.

**Comment by iclanton** [comment]
:shipit:

File: apps/rush/src/taskRunner/ProjectBuildTask.ts (hunk 3 above)
Comment (nickpape-msft) [inline apps/rush/src/taskRunner/ProjectBuildTask.ts]: we have a bug for it, but definitely out of scope for this PR

---
In reply to: [116883375](https://github.com/Microsoft/web-build-tools/pull/195#discussion_r116883375) [](ancestors = 116883375)

**Comment by nickpape-msft** [comment]
+1

---
In reply to: [301945319](https://github.com/Microsoft/web-build-tools/pull/195#issuecomment-301945319) [](ancestors = 301945319)

**Pull request merged**